src/wmi.c

```diff
@@ -0,0 +1,548 @@
+    params.cNamedArgs = 0;
+
+    params.rgvarg = args;
+    params.rgvarg[0].vt = VT_BSTR;
+    params.rgvarg[0].bstrVal = SysAllocString(name);
+
+    hr = wmi_invoke_method (varResult.pdispVal, L"Item", &params, &varResult);
+    SysFreeString(params.rgvarg[0].bstrVal);
+    if (hr != S_OK)
+        goto err;
+
+    hr = wmi_get_property (varResult.pdispVal, L"Value", value);
+    if (hr != S_OK)
+        goto err;
+
+    return (0);
+
+err:
+    // TODO: do proper error handling
+    switch (hr)
+    {
+    case WBEM_E_NOT_FOUND:
+        ERROR ("wmi error: Property %ls not found.", name);
+        break;
+    default:
+        ERROR ("wmi error: Unknown error 0x%x while fetching property %ls",
+              (unsigned) hr, name);
+        break;
+    }
+
+    return (-1);
+}
+
+void wmi_release(wmi_connection_t *connection)
+{
+    if (!connection)
+        return;
+
+    if (connection->dispatcher)
+        connection->dispatcher->lpVtbl->Release(connection->dispatcher);
+
+    CoUninitialize();
+
+    free (connection);
+}
+
+wmi_connection_t* wmi_connect()
+{
+    wmi_connection_t *connection = malloc (sizeof (wmi_connection_t));
+    HRESULT hr;
+    hr = CoInitializeEx(0, COINIT_MULTITHREADED);
+    if (hr != S_OK)
+        goto err;
+
+    BSTR str = SysAllocString(L"winmgmts:root\\cimv2");
+    hr = CoGetObject(str, NULL, &IID_IDispatch, (void**)&connection->dispatcher);
+    if (hr != S_OK)
+        goto err;
+
+    return connection;
+
+err:
+    ERROR("wmi error: Initialization failed. Error code: %x", (unsigned)hr);
+    wmi_release (connection);
+    return NULL;
+}
+
+static wmi_connection_t *wmi;
+
+static int wmi_init (void)
+{
+    wmi = wmi_connect();
+    return (0);
+}
+
+void plugin_instance_free (plugin_instance_t *pi)
+{
+    if (!pi) return;
+
+    free (pi->base_name);
+    LIST_FREE (pi->queries, wmi_query_free);
+}
+
+static int wmi_shutdown (void)
+{
+    LIST_FREE (plugin_instances_g, plugin_instance_free);
+    wmi_release (wmi);
+    return (0);
+}
+
+static void store (VARIANT *src, value_t *dst, int dst_type)
+{
+    switch (dst_type)
+    {
+    case DS_TYPE_GAUGE:
+        dst->gauge = variant_get_double (src);
+        break;
+
+    case DS_TYPE_DERIVE:
+        dst->derive = variant_get_int64 (src);
+        break;
+
+    case DS_TYPE_ABSOLUTE:
+        dst->absolute = variant_get_uint64 (src);
+        break;
+
+    case DS_TYPE_COUNTER:
+        dst->counter = variant_get_ull (src);
+        break;
+
+    default:
+        ERROR ("Destination type '%d' is not supported", dst_type);
+        break;
+    }
+}
+
+/* Find position of `name` in `ds */
+static int find_index_in_ds (const data_set_t *ds, const char *name)
+{
+    int i;
+    for (i = 0; i < ds->ds_num; i++)
+        if (strcmp (ds->ds[i].name, name) == 0)
+            return (i);
+    return (-1);
+}
+
+static void sanitize_string (char *s)
+{
+    int i;
+    for (i = 0; s[i]; i++)
+    {
+        if (isalnum (s[i]) || s[i] == '-')
```

Comment:
Can't you just make it:
if (!(isalnum (s[i]) || s[i] == '-'))
    s[i] = '_';
?